merge: treat nodata=None and np.nan in data

j08lue opened the pull request:
## The issue

I want to merge some `float32` files that have `nodata=None` and contain `NaN` for missing values.

However, `merge` does not support `np.nan` as a nodata flag, because it uses comparison to validate the nodata value:
https://github.com/mapbox/rasterio/blob/89538636ff92112848a888e314bea8f457e3a30b/rasterio/merge.py#L117

and to test whether the existing data (`region`) is empty:
https://github.com/mapbox/rasterio/blob/89538636ff92112848a888e314bea8f457e3a30b/rasterio/merge.py#L171

## Proposed solution

If the `nodata` value is `np.nan`, accept it as valid for floating point data and use `np.isnan` to check whether the `region` is empty.

Also, for floating point data, if `nodata` is `None`, set it to `np.nan` per default. This part is perhaps too auto-magic, one could also leave it up to the user to specify `nodata=np.nan` in the function call.

## Relevance

I do not know how common it is to use `np.nan` as implicit or explicit nodata value in floating point data. But QGIS, for example, handles this case out of the box.

j08lue commented:
I can provide demo data if you like.

mwtoews commented:
If a raster has `NaN` for missing values, then shouldn't `nodata == np.nan`? I interpret `nodata is None` as an unset NoData condition.

mwtoews commented:
One shortcoming I see is that `rio` cannot set NaN, but `gdal_edit.py` can:
```
$ rio edit-info --nodata NaN file.tif
Error: Invalid value for nodata: outside the range of the file's data type (float32).
$ gdal_edit.py -a_nodata NaN file.tif
# works
```

j08lue commented:
> If a raster has NaN for missing values, then shouldn't nodata == np.nan?

Other software like QGIS automagically masks out `np.nan`, even when `nodata is None`. But I can definitely remove that part, if you like, @mwtoews?

> One shortcoming I see is that rio cannot set NaN, but gdal_edit.py can:

I can fix that in `edit_info` in the same way as above: When validating the `nodata` value and encountering `np.nan` in `float` data, validation passes. Should I add that to this PR, @mwtoews?

sgillies commented:
I'm hesitant to make nan an implicit nodata value. I feel like it goes against the grain of numpy, and rasterio aims to be more consistent with numpy than with QGIS.

For example:

```python
>>> arr = np.array([1.0, np.nan])
>>> arr
array([  1.,  nan])
>>> arr.mean()
nan
```

If you have `np.nan` in an array, it isn't ignored.

j08lue commented:
I removed the three lines (auto-magic `None` to `nan` conversion) and added support for setting `--nodata NaN` in `rio edit-info`.

You can verify that it works with

```bash
rio edit-info --unset-nodata /path/to/float_file.tif
rio info --nodata /path/to/float_file.tif  # returns nothing but whitespace
rio edit-info --nodata NaN /path/to/float_file.tif
rio info --nodata /path/to/float_file.tif  # returns nan
```

As far as I can see, we neither have tests for `rio edit-info` nor for `rasterio.merge`? We only have a test for [`rio merge`](https://github.com/mapbox/rasterio/blob/master/tests/test_rio_merge.py). I would be happy to contribute here, including coverage for the NaN part, but perhaps on another PR.

j08lue commented:
I added a test for `rio edit-info`. 

Do you need me to make one for the special case of `merge(..., nodata=np.nan)`, too? I could add that to [`test_rio_merge`](https://github.com/mapbox/rasterio/blob/master/tests/test_rio_merge.py), generating fake data in the same manner as [`test_data_dir1`](https://github.com/mapbox/rasterio/blob/master/tests/test_rio_merge.py#L27-L79) - but maybe a bit much extra code for such a small feature?

sgillies commented:
Thanks @j08lue !
